### PR TITLE
Add REP process definition and template

### PR DESCRIPTION
**Pull request by shenglol**
- Updated README.md to define the Bicep REP workflow
- Added a REP template

References:
- https://github.com/PowerShell/PowerShell-RFC
- https://github.com/rust-lang/rfcs
- https://github.com/reactjs/rfcs

### CODE REVIEW

File: 0000-rep-template.md [hunk 1]
```diff
@@ -0,0 +1,45 @@
+---
+REP Number: <Fill me in with a four-digit unique incrementing number, matching the pull request number; update AFTER REP acceptance.>
+Author: <Author's First Name and Last Name.>
```
Comment (StephenWeatherford) [inline 0000-rep-template.md:3]: not github username?

File: 0000-rep-template.md [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+- Example 1
+```bicep
+```
+- Example 2
+```bicep
+```
+
+## Drawbacks
+
+Discuss any potential drawbacks associated with this approach. Why might this *not* be a suitable choice? Consider implementation costs, integration with existing and planned features, and the migration cost for Bicep users. Clearly identify if it constitutes a breaking change.
```
Comment (majastrz) [inline 0000-rep-template.md:37]: Should we have some sort of title modifier or text or tag somewhere that makes it more obvious that we're dealing with a breaking changes REP?

File: 0000-rep-template.md [hunk 3]
```diff
@@ -0,0 +1,45 @@
+---
+REP Number: <Fill me in with a four-digit unique incrementing number, matching the pull request number; update AFTER REP acceptance.>
+Author: <Author's First Name and Last Name.>
+Start Date: <Today's Date, YYYY-MM-DD>
+Bicep Issue Number: <Fill if the REP originates from a Bicep issue; otherwise, delete.>
```
Comment (majastrz) [inline 0000-rep-template.md:5]: Could be multiple issues in some cases. Maybe we just need to add " (s)" to the title?

File: README.md [hunk 4]
```diff
@@ -1,33 +1,81 @@
...
 
-As the maintainer of this project, please make a few updates:
+## When to Follow the Process
 
-- Improving this README.MD file to provide a great experience
-- Updating SUPPORT.MD with content about this project's support experience
-- Understanding the security reporting process in SECURITY.MD
-- Remove this section from the README
+Consider employing this process if you intend to make substantial changes to the Bicep programming language that would affect user experience. Example where the use of a REP is warranted include:
+- *Introduction of New Features*: If the proposed change involves the addition of a new syntax kind or a new ARM template function, particularly if it necessitates the use of a feature flag.
```
Comment (alex-frankel) [inline README.md:8]: "feature" is pretty broad - do we specifically mean a syntax change?

File: README.md [hunk 5]
```diff
@@ -1,33 +1,81 @@
...
+### Draft
+
+This is the initial stage of a Bicep REP, signaling it as a work-in-progress. To create a draft REP, follow these steps:
+- Copy `0000-rep-template.md` to `active/0000-my-proposed-feature` (replace `my-proposed-feature` with the actual feature name). Avoid assigning an REP number at this point.
+- Complete the REP template.
+    - Optionally, If the REP is an enhancement of an existing REP that does not introduce any breaking changes, the author should fill in the `Enhances` field to reference the REP that is being improved.
+    - Optionally, if the REP deprecates an existing REP, the author should fill in the `Deprecates` field to reference the REP that is being deprecated.
+- Submit a pull request. During the pull request phase, the REP will undergo design feedback from both Bicep core maintainers and the broader community. The author should be prepared to make revisions based on this feedback.
+- Build consensus and integrate feedback.
+- When the Bicep core maintainers feel confident about the design and have reached a consensus, they should proceed to vote on whether to accept or reject the REP.
```
[the remark this answers is not in the document]
Comment (majastrz) [reply README.md:33]: Also do we want to have a more light-weight process for user suggestions before (or if) they turn into REPs?

File: README.md [hunk 6]
```diff
@@ -1,33 +1,81 @@
...
+While members of the Bicep community are allowed to contribute by authoring new REPs and providing feedback, it's important to note that, in practice, Bicep REPs are typically submitted by the Bicep core maintainers. This submission follows an exclusive design and discussion process among the maintainers. The author of a Bicep REP is not only responsible for the initial design but also for implementing the proposed feature and overseeing the REP's lifecycle.
 
-This project has adopted the [Microsoft Open Source Code of Conduct](https://opensource.microsoft.com/codeofconduct/).
-For more information see the [Code of Conduct FAQ](https://opensource.microsoft.com/codeofconduct/faq/) or
-contact [opencode@microsoft.com](mailto:opencode@microsoft.com) with any additional questions or comments.
+Therefore, it is generally not recommended for community members to independently initiate the creation of REPs. Instead, we encourage active collaboration and engagement within the main Bicep repository. This involves participating in issue discussions and upvoting feature requests, which can prompt the Bicep core maintainers to initiate the creation of Bicep REPs.
```
Comment (alex-frankel) [inline README.md:18]: should we be more specific thatm in most cases, starting with an issue for a feature proposal is more appropriate?

File: 0000-rep-template.md (hunk 1 above)
Comment (shenglol) [reply 0000-rep-template.md:3]: Using both makes sense.

File: 0000-rep-template.md [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+```bicep
+```
+- Example 2
+```bicep
+```
+
+## Drawbacks
+
+Discuss any potential drawbacks associated with this approach. Why might this *not* be a suitable choice? Consider implementation costs, integration with existing and planned features, and the migration cost for Bicep users. Clearly identify if it constitutes a breaking change.
+
```
Comment (asilverman) [inline 0000-rep-template.md:38]: Should we add an Out of Scope:
> [Out of scope](https://gist.github.com/rynowak/5a39b92997546216f957f0aa2dc4b32b#out-of-scope)
Describe the important issues or concerns NOT addressed by this proposal. 1 paragraph each (can be short)

Often the "out of scope" items are refer to the current state. We might already have planned work to address the "out of scope" items, mention it if that's the case".

File: README.md [hunk 8]
```diff
@@ -1,33 +1,81 @@
...
+Once a REP is reviewed and approved, it is considered active. The author should take the following steps to maintain the active state of the REP:
+- Replace `0000` in the REP file name with the PR number.
+_ Merge the PR so that the REP file will appear in the active folder in the main branch.
+- Proceed with the implementation of the proposed feature.
+- Optionally, consider putting the feature behind a preview feature flag. This would give the author the opportunity to easily make revisions and breaking changes.
+- As the author works towards a final implementation, submit PRs to the bicep-reps repository to keep the active REP in sync with the implementation.
+
+### Final
+
+REPs in the Final state are considered fully complete and implemented without any preview feature flag. Any proposed changes should be made through a new REP which enhances or deprecates the final REP. To convert an active REP to a final, follow these steps:
```
Comment (shenglol) [inline README.md:46]: Make it clear that all unresolved questions are resolved.

File: README.md [hunk 9]
```diff
@@ -1,33 +1,81 @@
...
-Contributor License Agreement (CLA) declaring that you have the right to, and actually do, grant us
-the rights to use your contribution. For details, visit https://cla.opensource.microsoft.com.
+## Who Should Create Bicep REPs
 
-When you submit a pull request, a CLA bot will automatically determine whether you need to provide
-a CLA and decorate the PR appropriately (e.g., status check, comment). Simply follow the instructions
-provided by the bot. You will only need to do this once across all repos using our CLA.
+While members of the Bicep community are allowed to contribute by authoring new REPs and providing feedback, it's important to note that, in practice, Bicep REPs are typically submitted by the Bicep core maintainers. This submission follows an exclusive design and discussion process among the maintainers. The author of a Bicep REP is not only responsible for the initial design but also for implementing the proposed feature and overseeing the REP's lifecycle.
 
-This project has adopted the [Microsoft Open Source Code of Conduct](https://opensource.microsoft.com/codeofconduct/).
```
Comment (StephenWeatherford) [inline README.md:23]: Code of conduct section might need to be here still...?

File: README.md [hunk 10]
```diff
@@ -1,33 +1,81 @@
...
+Refer to the state diagram below for a visual representation of the Bicep REP lifecycle.
+
+```mermaid
+stateDiagram-v2
+    state Review <<choice>>
+    [*] --> Draft: Open a PR
+    Draft --> Review: Review the PR
+    Review --> Active: PR accepted
+    Review --> Rejected: PR rejected
+    Active --> Active: Create new PRs to\n make revision
```
Comment (majastrz) [inline README.md:76]: Do we need an Experimental REP state similar to how PowerShell RFCs handle experimental features?

File: README.md [hunk 11]
```diff
@@ -1,33 +1,81 @@
...
+While members of the Bicep community are allowed to contribute by authoring new REPs and providing feedback, it's important to note that, in practice, Bicep REPs are typically submitted by the Bicep core maintainers. This submission follows an exclusive design and discussion process among the maintainers. The author of a Bicep REP is not only responsible for the initial design but also for implementing the proposed feature and overseeing the REP's lifecycle.
 
-This project has adopted the [Microsoft Open Source Code of Conduct](https://opensource.microsoft.com/codeofconduct/).
-For more information see the [Code of Conduct FAQ](https://opensource.microsoft.com/codeofconduct/faq/) or
-contact [opencode@microsoft.com](mailto:opencode@microsoft.com) with any additional questions or comments.
+Therefore, it is generally not recommended for community members to independently initiate the creation of REPs. Instead, we encourage active collaboration and engagement within the main Bicep repository. This involves participating in issue discussions and upvoting feature requests, which can prompt the Bicep core maintainers to initiate the creation of Bicep REPs.
 
-## Trademarks
```
Comment (StephenWeatherford) [inline README.md:27]: ditto

File: 0000-rep-template.md [hunk 12]
```diff
@@ -0,0 +1,45 @@
+---
+REP Number: <Fill me in with a four-digit unique incrementing number, matching the pull request number; update AFTER REP acceptance.>
```
Comment (StephenWeatherford) [inline 0000-rep-template.md:2]:
```suggestion
REP Number: 0000 <Replace me in with the pull request number update AFTER REP acceptance.>
```

Comment (StephenWeatherford) [inline 0000-rep-template.md:2]: This is confusing.  This should be the pull request number?  If so, "unique incrementing number" doesn't make sense.  So this should be left blank until it's accepted?  Can a REP move from draft immediately to inactive?

File: 0000-rep-template.md [hunk 13]
```diff
@@ -0,0 +1,45 @@
...
+
+# Title - replace me with feature name
+
+## Summary
+
+Provide a brief paragraph explaining the feature.
+
+## Motivation
+
+Outline the reasons behind implementing this feature, the use cases it supports, and the expected outcomes.
```
Comment (shenglol) [inline 0000-rep-template.md:20]: Add [Terms and definitions](https://github.com/radius-project/design-notes/blob/main/template/YYYY-MM-design-template.md#terms-and-definitions).

File: README.md [hunk 14]
```diff
@@ -1,33 +1,81 @@
...
+    state Review <<choice>>
+    [*] --> Draft: Open a PR
+    Draft --> Review: Review the PR
+    Review --> Active: PR accepted
+    Review --> Rejected: PR rejected
+    Active --> Active: Create new PRs to\n make revision
+    Active --> Final: Feature fully implemented
+    Active --> Inactive: Not enough feedback\n has been received to\n finalize the feature
+    Final --> Deprecated: REP superceded by a new REP
+    Rejected --> [*]
```
Comment (StephenWeatherford) [inline README.md:80]: Are there times we might want to keep the rejected proposals for records?  Rejecting and deleting the PR and branch are permanent.  Maybe optionally move to a rejected folder?

File: README.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (shenglol) [reply README.md:33]: Suggested final approvers: @majastrz @anthony-c-martin @alex-frankel .

[the remark this answers is not in the document]
Comment (shenglol) [reply README.md:33]: We can use Bicep discussion meetings to review REPs.

[the remark this answers is not in the document]
Comment (shenglol) [reply README.md:33]: Number of approvals does not matter.

File: README.md [hunk 15]
```diff
@@ -1,33 +1,81 @@
...
+- As the author works towards a final implementation, submit PRs to the bicep-reps repository to keep the active REP in sync with the implementation.
+
+### Final
+
+REPs in the Final state are considered fully complete and implemented without any preview feature flag. Any proposed changes should be made through a new REP which enhances or deprecates the final REP. To convert an active REP to a final, follow these steps:
+- Ensure that the active REP is fully implemented and well tested, and there will be no breaking changes to the customer facing API contract.
+- Submit a PR to move the REP from the active folder to the final folder.
+    - Optionally, as part of the PR change, if the REP marked as final enhances an existing final REP, include in the existing REP a `Enhanced By` field to reference the new REP.
+
+### Inactive
```
Comment (shenglol) [inline README.md:51]: Inactive => Deferred

File: README.md (hunk 14 above)
Comment (shenglol) [reply README.md:80]: Based on the discussion, we'd like to use closed PRs to track them to reduce mantainance overheads.

File: README.md (hunk 10 above)
Comment (shenglol) [reply README.md:76]: My expectation is that most REP features are likely to be initially implemented as experimental, so I'm not sure if it's worth maintaining two active states, one for non-experimental features and one for experimental features. What do you think?

File: 0000-rep-template.md (hunk 12 above)
Comment (shenglol) [reply 0000-rep-template.md:2]: PR numbers are unique incrementing numbers, but I guess it is a bit confusing to explicitly mention it here. The idea is to left it blank until it's accepted and before merged. Will update this to reflect that.

Comment (shenglol) [reply 0000-rep-template.md:2]: > Can a REP move from draft immediately to inactive?

No, only a PR that tries to add a REP file to the `/active` folder is considered a draft.

File: 0000-rep-template.md (hunk 2 above)
Comment (shenglol) [reply 0000-rep-template.md]: Update the REP template title to handle this.

File: 0000-rep-template.md (hunk 3 above)
Comment (shenglol) [reply 0000-rep-template.md]: Fixed.

File: README.md (hunk 4 above)
Comment (shenglol) [reply README.md]: Improved this section and added a few more examples.

File: README.md (hunk 6 above)
Comment (shenglol) [reply README.md]: Yeah makes sense. Incorporated that into this section.

File: 0000-rep-template.md (hunk 7 above)
Comment (shenglol) [reply 0000-rep-template.md:51]: Added an Out of scope section.

File: README.md [hunk 16]
```diff
@@ -1,33 +1,81 @@
...
+This is the initial stage of a Bicep REP, signaling it as a work-in-progress. To create a draft REP, follow these steps:
+- Copy `0000-rep-template.md` to `active/0000-my-proposed-feature` (replace `my-proposed-feature` with the actual feature name). Avoid assigning an REP number at this point.
+- Complete the REP template.
+    - Optionally, If the REP is an enhancement of an existing REP that does not introduce any breaking changes, the author should fill in the `Enhances` field to reference the REP that is being improved.
+    - Optionally, if the REP deprecates an existing REP, the author should fill in the `Deprecates` field to reference the REP that is being deprecated.
+- Submit a pull request. During the pull request phase, the REP will undergo design feedback from both Bicep core maintainers and the broader community. The author should be prepared to make revisions based on this feedback.
+- Build consensus and integrate feedback.
+- When the Bicep core maintainers feel confident about the design and have reached a consensus, they should proceed to vote on whether to accept or reject the REP.
+
+### Active
```
[the remark this answers is not in the document]
Comment (shenglol) [reply README.md:42]: Yeah, added a blockquote to mention that.

File: 0000-rep-template.md [hunk 17]
```diff
@@ -0,0 +1,69 @@
...
+> Describe how you will deliver your feature. Does it need to be an experimental feature? If the feature involves backend changes, dose it need to be guarded by an ARM feature flag?
+
+## Unresolved questions
+
+> - What aspects of the design do you expect to resolve through the REP process before merging?
+> - What parts of the design do you anticipate resolving through feature implementation before stabilization?
+
+## Out of scope
+
+> - What related issues are considered out of scope for this REP but could be addressed independently in the future?
```
[the remark this answers is not in the document]
Comment (shenglol) [reply 0000-rep-template.md:67]: Good suggestion!

**Review by anthony-c-martin — APPROVED** [review]

**Pull request merged**